Chat: allow large file with range in @-mentions

abeatrix opened the pull request:
CLOSE: https://github.com/sourcegraph/cody/issues/3589#issuecomment-2026647366
PART OF:  https://github.com/sourcegraph/cody-issues/issues/9

### Demo

https://github.com/sourcegraph/cody/assets/68532117/8e0ee5e8-40b4-4b1a-b07b-7420af3f5183

This pull request updates the prompt editor to allow large files to be added to the context when they have a specified range. Previously, large files were not allowed to be added to the context file list. Now, large files with ranges can be included in the context.

##Changes

Update contextItemMentionNodeDisplayText in ContextItemMentionNode.ts:

- Remove the condition that prevents large files from being added to the context.
- Always display the file path and range (if available) for files.

Update Item component in OptionsList.tsx:

- Modify the condition for showing the "File too large" warning to only display when the file is large and doesn't have a range specified.

Update MentionsPlugin in atMentions.tsx:

- Differentiate between file selections and other selections.
- For file selections, replace the node with the selected file path and range (if available).
- If the selected file is large and doesn't have a range, append a `:` after the file path to allow adding line range.
- Create a mention node for the selected item and insert it into the document.

## Limitation

Users will not see the large file warning on large file with range input because we will need to get the range numbers first before we can check the token usage, which should be part of the works when we start working on improving our context window logic.

Tracking issue: https://github.com/sourcegraph/cody-issues/issues/9


## Test plan

<!-- Required. See https://sourcegraph.com/docs/dev/background-information/testing_principles. -->

Search for a large file via @-mentions:

 ![image](https://github.com/sourcegraph/cody/assets/68532117/d7ceb432-e747-4b90-8dde-ff6db39e1430)

Press tab, verify `:` is added to the end of the file path. And then add file range:

 ![image](https://github.com/sourcegraph/cody/assets/68532117/ba8e4068-e710-4b0d-a3f0-6e11ad3a737a)

Hit tab / space again, and the large file is now added as a context item token:

![image](https://github.com/sourcegraph/cody/assets/68532117/f19c4391-fe91-4484-9f1b-8ddad7d46ef3)

Verify that selecting files that are not too large should automatically add to input box as token.

kalanchan reviewed (approved):
looks good! tested this on your branch and works 

edit: uploading proper video

kalanchan reviewed (dismissed):
The @ file with line numbers still got excluded in the context, did I test this correctly? 

also, it feels a bit weird to backspace after hitting tab/enter to enter in the line numbers, would it be better for large files to not have that empty space and let the user start selecting the line numbers right afterwards?

![CleanShot 2024-03-29 at 14 09 51](https://github.com/sourcegraph/cody/assets/51868853/8da3544b-8c09-4131-ba58-bf52f5acc65d)


https://github.com/sourcegraph/cody/assets/51868853/d89ca677-80bc-4bd7-b116-41609b743095

abeatrix commented:
> The @ file with line numbers still got excluded in the context, did I test this correctly?
> 
> also, it feels a bit weird to backspace after hitting tab/enter to enter in the line numbers, would it be better for large files to not have that empty space and let the user start selecting the line numbers right afterwards?
> 
> ![CleanShot 2024-03-29 at 14 09 51](https://private-user-images.githubusercontent.com/51868853/318138863-8da3544b-8c09-4131-ba58-bf52f5acc65d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE3NDk1NjQsIm5iZiI6MTcxMTc0OTI2NCwicGF0aCI6Ii81MTg2ODg1My8zMTgxMzg4NjMtOGRhMzU0NGItOGMwOS00MTMxLWJhNTgtYmY1MmY1YWNjNjVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI5VDIxNTQyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNlMzI2MDAxYjI4YmMzODM5OGMyZTcwYzEwNjA5NGIyOGE1NTlmMzMxNzg5NzU2OTE2MWJkNDE2MTU5NzgzYTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AAfNQOR0tDDLvtiyJ0GrLNi5UwJXOFAIhj-AdFh3860)
> 
>  large-file-excluded.mp4



> The @ file with line numbers still got excluded in the context, did I test this correctly?
> 
> also, it feels a bit weird to backspace after hitting tab/enter to enter in the line numbers, would it be better for large files to not have that empty space and let the user start selecting the line numbers right afterwards?
> 
> ![CleanShot 2024-03-29 at 14 09 51](https://private-user-images.githubusercontent.com/51868853/318138863-8da3544b-8c09-4131-ba58-bf52f5acc65d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE3NDk1NjQsIm5iZiI6MTcxMTc0OTI2NCwicGF0aCI6Ii81MTg2ODg1My8zMTgxMzg4NjMtOGRhMzU0NGItOGMwOS00MTMxLWJhNTgtYmY1MmY1YWNjNjVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI5VDIxNTQyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNlMzI2MDAxYjI4YmMzODM5OGMyZTcwYzEwNjA5NGIyOGE1NTlmMzMxNzg5NzU2OTE2MWJkNDE2MTU5NzgzYTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AAfNQOR0tDDLvtiyJ0GrLNi5UwJXOFAIhj-AdFh3860)
> 
>  large-file-excluded.mp4

I don't think that's related to my changes in this PR, since we can see the file was appended to the chat correctly, but got excluded by the chat context window since i suspect it's because you are using the smaller claude 3 model? Can you try Claude 3 Opus?

Mentioned in the Limitation section about more works need to be on the range for large files as well, which we can look into when we start the works on context windows?

kalanchan commented:
I think the current limitation is a bit confusing for the user, I don't know who to trust if I'm able to @ file a large file with range but the UI says it got excluded from context 😓 

@toolmantim @chillatom what do you all think? I think we need to address the limitations section either in the same PR or a follow up right afterwards

abeatrix commented:
> I don't know who to trust if I'm able to @ file a large file with range but the UI says it got excluded from context 

It was part of @toolmantim 's design and the engineering work on "We should not silently exclude a file that has been explicitly @ mentioned" 😅 We also have users asking to enable to @ line numbers again after we have disabled it on large files: https://github.com/sourcegraph/cody/pull/3523

@toolmantim  @chillatom @kalanchan I'll need more information on how we want the context window works before we can address this limitation too. I've sent you all an invite for next week so we can discuss live!

Issue for the warning title: https://github.com/sourcegraph/cody/issues/3620
I believe @chillatom has also created a tracking issue for improving the context window.

abeatrix commented:
@kalanchan You were right there was a bug with the display logic where used @-mention items are showing up as excluded when the file was originally too large. That's why in your screenshot Cody was still able to answer the question. 

I've fixed that so that all used files shouldn't be marked as large file:
![image](https://github.com/sourcegraph/cody/assets/68532117/527f1bb8-d4fa-4e9a-9d65-2428e3a38418)

Also updated to include `:1-10` when user try to include a large file to give them a head start:
![image](https://github.com/sourcegraph/cody/assets/68532117/c5a0deb4-e9c2-40e4-8f7a-2915ef4c7164)

abeatrix commented:
@toolmantim @chillatom updated the PR so that it matches the behavior we discussed earlier for Wednesday release: selecting a large file from the option list will add the file path to the input as text so that users can choose to add range after the auto-filled file path or remove it. Selecting files that are not large file will automatically turn it into a token. 

Follow-up from this would be allowing users to add selection from editor as @-mentions via right click menu. 

Let me know if I'm missing something here, thanks!

toolmantim reviewed (commented):
Looking good! I found a few things:

* Items shouldn't look disabled now they're clickable. ![CleanShot 2024-04-02 at 22 43 24@2x](https://github.com/sourcegraph/cody/assets/153/1a4496f7-c00f-4337-add2-dd45a55456c9) Can we bring back up the opacity and cursor etc?
* When you select an item, can we complete the path, add `:` to the end, make sure it's the only item in the menu, and give them some hint for how to type the range? At the moment you need to manually type the `:` and then it starts doing a line-range to all items: 
![CleanShot 2024-04-02 at 22 43 07@2x](https://github.com/sourcegraph/cody/assets/153/0fc0186f-2c1b-443a-85f8-43889acd5a5f)
* When you add `:` can we replace the "File is too large ..." file subtext with something more helpful like "Type a line range to include, e.g. 1-42": ![CleanShot 2024-04-02 at 22 56 46@2x](https://github.com/sourcegraph/cody/assets/153/c4691c96-b427-4dda-ad64-04638e65e536)

* Let's remove the "File will be excluded if context range exceeds limit" message. Could we implement a maximum number of lines you can do, and only if they exceed it we show the message, rather than warn everyone ahead of time?
* If you type a line range (e.g. `:1-42`) and remove the lines, leaving only `:`, it seems to show an old cached value: ![CleanShot 2024-04-02 at 22 49 20@2x](https://github.com/sourcegraph/cody/assets/153/572a9cee-3322-4f90-93db-13da3789dc5f)  ![CleanShot 2024-04-02 at 22 52 34@2x](https://github.com/sourcegraph/cody/assets/153/095808a0-22c3-447e-8b0d-67190a46da6a)

umpox commented:
Agree with Tim's comments about the disabled text.

Noticed that this doesn't work with the Edit input right now, is it supposed to?

<img width="689" alt="image" src="https://github.com/sourcegraph/cody/assets/9516420/2346e4dd-b6db-439a-9471-c278380d9285">

We can do this in a follow up but we should not show the same warning if so:
<img width="691" alt="image" src="https://github.com/sourcegraph/cody/assets/9516420/e685875c-3f15-4509-ad07-82ba015f4e0d">

abeatrix commented:
@toolmantim @umpox Thanks for the review! I've updated the PR to address the following changes:

- [x] Items shouldn't look disabled now they're clickable. we bring back up the opacity and cursor etc?
  - Opacity updated from 0.5 to 0.8. Let me know if you think we should just remove it.
  ![image](https://github.com/sourcegraph/cody/assets/68532117/d7ceb432-e747-4b90-8dde-ff6db39e1430)

- [x] When you select an item, can we complete the path, add : to the end, 
- [x] make sure it's the only item in the menu
- [x] give them some hint for how to type the range
  - Updated title to “Type a line range to include, e.g. 1-42”
- [x] When you add : can we replace the "File is too large ..." file subtext with something more helpful like "Type a line range to include, e.g. 1-42": 
  - Updated title to “Type a line range to include, e.g. 1-42”
  ![image](https://github.com/sourcegraph/cody/assets/68532117/ba8e4068-e710-4b0d-a3f0-6e11ad3a737a)
- [x] Let's remove the "File will be excluded if context range exceeds limit" message.
- [x] If you type a line range (e.g. :1-42) and remove the lines, leaving only :, it seems to show an old cached value:
- [x] Let's remove the "File will be excluded if context range exceeds limit" message
- [ ]  Could we implement a maximum number of lines you can do, and only if they exceed it we show the message, rather than warn everyone ahead of time?
  - This would be in the scope for the works we are starting in this sprint right? To find the right number?
  
> Noticed that this doesn't work with the Edit input right now, is it supposed to?
@umpox This is only updating the UI for chat, so I don't suppose the changes in this PR will work with Edit which uses a different UI. Will create an issue for Edit


### Demo

https://github.com/sourcegraph/cody/assets/68532117/8e0ee5e8-40b4-4b1a-b07b-7420af3f5183

sqs commented on vscode/CHANGELOG.md:
```diff
@@ -18,7 +18,7 @@ This is a log of all notable changes to Cody for VS Code. [Unreleased] changes a
 ### Fixed
 
-- Chat: Large file cannot be added via @-mention. [pull/3531](https://github.com/sourcegraph/cody/pull/3531)
+- Large files cannot be added via @-mention without specifying a line range. [pull/3531](https://github.com/sourcegraph/cody/pull/3531) & [pull/3585](https://github.com/sourcegraph/cody/pull/3585)
```
```suggestion
- Chat: Fixed issue where large files could not be added via @-mention. You can now @-mention line ranges within large files. [pull/3531](https://github.com/sourcegraph/cody/pull/3531) & [pull/3585](https://github.com/sourcegraph/cody/pull/3585)
```

sqs commented on vscode/src/edit/input/get-input.ts:
```diff
@@ -413,7 +413,9 @@ export const getInput = async (
                         description: shortLabel ? key : undefined,
                         detail:
-                            'isTooLarge' in item && item.isTooLarge ? FILE_TOO_LARGE_LABEL : undefined,
+                            'isTooLarge' in item && item.isTooLarge
```
```suggestion
                            item.type === 'file && item.isTooLarge
```

sqs commented on vscode/webviews/promptEditor/plugins/atMentions/OptionsList.tsx:
```diff
@@ -49,7 +50,9 @@ export const OptionsList: FunctionComponent<
                                   (mentionQuery.text.length < 3 ? NO_SYMBOL_MATCHES_HELP_LABEL : '')
                           : options.length > 0
-                              ? FILE_HELP_LABEL
+                              ? isValideLineRangeQuery(query)
```
```suggestion
                              ? isValidLineRangeQuery(query)
```

sqs reviewed (approved):
Code LGTM with comments, have not tried this yet but I see other people have

abeatrix commented on vscode/src/edit/input/get-input.ts:
```diff
@@ -413,7 +413,9 @@ export const getInput = async (
                         description: shortLabel ? key : undefined,
                         detail:
-                            'isTooLarge' in item && item.isTooLarge ? FILE_TOO_LARGE_LABEL : undefined,
+                            item.type === 'file && item.isTooLarge
```
ty!

toolmantim commented on vscode/webviews/promptEditor/plugins/atMentions/OptionsList.module.css:
```diff
@@ -72,7 +72,7 @@
     background-color: var(--vscode-list-hoverBackground);
 }
 .option-item.disabled {
-    opacity: 0.5;
```
Yeah, let's not change the opacity at all, as we have the label and don't want it to feel disabled. We could create a custom icon if we really wanted, but I think the label is good.

We could add a right chevron to the right hand side ">" to show that you don't be in fact selecting the item, but "going into it"

abeatrix replied:
@toolmantim Thanks! removing opacity in https://github.com/sourcegraph/cody/pull/3665

hpn1975 commented:
I am sorry.. but for a paid member keep getting this error is annoying... the file has only 75 lines of code.
![image](https://github.com/sourcegraph/cody/assets/25471567/da05d4bd-24af-458d-924d-0d4ff832c46a)
I am not sure what you guys are fixing or what is fixed.  And having to enter the line number is so annoying..
